### PR TITLE
CVSL-242 offender image upload

### PR DESCRIPTION
**Pull request by timharrison-moj**
This PR:
* Fixes an existing bug when retrieving the image by nomisID when multiple bookings exists could return an old, even inactive image. Retrieve by bookingId was returning the correct, latest image.
* Adds a new endpoint `PUT /api/images/offender/{prisonerNumber}`
* Add a method to the ImageService to receive, convert, scale and apply full size and thumbnail images to the repository
* Adjusts the OffenderImage entity to use the OFFENDER_IMAGE_ID sequence and use @Lob type columns for byte[] images.
* Sets the previously active image to active-false.
* Tests for service & resource integration to verify.
* README changes with instructions for converting / uploading an image.
* I've left the swagger docs hidden for now - we can try in dev and when happy turn this on. 
* It requires ROLE_IMAGE_UPLOAD to access.

image

### CODE REVIEW

File: src/test/resources/uk/gov/justice/hmpps/prison/api/resource/impl/offender_sentences.json [hunk 1]
```diff
@@ -95,8 +95,7 @@
   },
   "dateOfBirth": "1970-03-01",
   "agencyLocationDesc": "LEEDS",
-  "internalLocationDesc": "A-1-5",
-  "facialImageId": -9
+  "internalLocationDesc": "A-1-5"
 }, {
```
Comment (timharrison-moj) [inline src/test/resources/uk/gov/justice/hmpps/prison/api/resource/impl/offender_sentences.json:99]: Necessary to remove the `facialImageId` check here - as this offender was also used for the image upload integration tests, and depending on the order they run in, may or may not be -9. Removing it makes this work in either order.

File: README.MD [hunk 2]
```diff
@@ -216,3 +216,29 @@ There is an alert in Application Insights called `Elite2-api - Inactivity alert`
 
 If the alert fires then look for any recent releases of prison-api that may have introduced a problem. If not then check in the `nomisapi-prod` Kubernetes cluster looking for any errors - look at events, ingresses, certificates etc. Finally it could be some kind of network issue - check for problems in Azure AKS.
 
+
+### Offender images
+
+* Offender images can be set using the endpoint PUT /api/images//offender/{offenderNo}
+* This requires a token with the role ROLE_SYSTEM_USER
```
Comment (Mjwillis) [inline README.MD:223]: Change the role to ROLE_UPLOAD_IMAGE

File: README.MD [hunk 3]
```diff
@@ -216,3 +216,29 @@ There is an alert in Application Insights called `Elite2-api - Inactivity alert`
...
+* The body of the request should be a base64-encoded JPEG file
+* The optional query parameter ?fullSizeImage=true can be used if this is a full-sized image.
```
Comment (Mjwillis) [inline README.MD:225]: Nomis creates both full size and thumbnail by default - so the image uploaded is full size (assumed) and then a thumbnail is generated and stored as well

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java [hunk 4]
```diff
@@ -66,4 +72,19 @@
     public ImageDetail getImage(@PathVariable("imageId") @ApiParam(value = "The image id of offender", required = true) final Long imageId) {
         return imageService.findImageDetail(imageId);
     }
+
+    @ApiResponses({
+        @ApiResponse(code = 200, message = "OK", response = ImageDetail.class),
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:77]: Not needed

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java [hunk 5]
```diff
@@ -66,4 +72,19 @@
...
+        @ApiResponse(code = 400, message = "Invalid request.", response = ErrorResponse.class),
+        @ApiResponse(code = 404, message = "The offender number could not be found or has no bookings.", response = ErrorResponse.class),
+        @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class)})
+    @ApiOperation(value = "Upload a new image for an offender.", notes = "Image data is base64-encoded JPEG content. Requires ROLE_SYSTEM_USER.", nickname = "putImage", hidden = true)
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:81]: Suggest new role ROLE_IMAGE_UPLOAD

Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:81]: We are trying to deprecate SYSTEM_USER

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java [hunk 6]
```diff
@@ -66,4 +72,19 @@
...
+    @PutMapping("/offender/{offenderNo}")
+    public ImageDetail putImage(
+        @Pattern(regexp = "^[A-Z]\\d{4}[A-Z]{2}$", message = "Offender Number format incorrect") @PathVariable("offenderNo") @ApiParam(value = "The offender number relating to this image.", required = true) final String offenderNo,
+        @RequestParam(value = "fullSizeImage", defaultValue = "false") @ApiParam(value = "Set to true if this is a full size image", defaultValue = "false") final boolean fullSizeImage,
+        @RequestBody @ApiParam(value = "The base64-encoded string of the JPEG image data.", required = true) final String imageData
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:86]: could this be byte[] instead of String?

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java [hunk 7]
```diff
@@ -1,6 +1,5 @@
 package uk.gov.justice.hmpps.prison.repository.jpa.repository;
 
-
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java:3]: Redundant

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java [hunk 8]
```diff
@@ -47,5 +51,41 @@ public ImageDetail findImageDetail(final Long imageId) {
             .map(i -> fullSizeImage ? i.getFullSizeImage() : i.getThumbnailImage());
     }
 
+    @PreAuthorize("hasRole('SYSTEM_USER')")
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:54]: Change to IMAGE_UPLOAD

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java [hunk 9]
```diff
@@ -47,5 +51,41 @@ public ImageDetail findImageDetail(final Long imageId) {
...
+    @HasWriteScope
+    @Transactional
+    public ImageDetail putImageForOffender(final String offenderNumber, final boolean fullSizeImage,  final String imageData) {
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:57]: This should have `@VerifyOffenderAccess(overrideRole = 'ROLE_IMAGE_UPLOAD')`

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java [hunk 10]
```diff
@@ -47,5 +51,41 @@ public ImageDetail findImageDetail(final Long imageId) {
...
+        // Check that the offender number exists
+        List<Offender> offenderList = offenderRepository.findByNomsId(offenderNumber);
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:59]: Use `offenderRepository.findOffenderByNomsId`

Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:59]: JPA version

Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:59]: `offenderRepository.findOffenderByNomsId(offenderNo)            .orElseThrow(EntityNotFoundException.withMessage(format("No prisoner found for prisoner number %s", offenderNo)));`

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java [hunk 11]
```diff
@@ -47,5 +51,41 @@ public ImageDetail findImageDetail(final Long imageId) {
...
+        if (offenderList.isEmpty()) {
+            throw EntityNotFoundException.withId(offenderNumber);
+        }
+
+        // Check that there is a booking present and disallow when not present. Must there be a booking?
+        var latestBooking = offenderList.get(0).getLatestBooking();
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:65]: See above ` Optional<OffenderBooking> getLatestBooking() `  instead on JPA Offender entity

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java [hunk 12]
```diff
@@ -47,5 +51,41 @@ public ImageDetail findImageDetail(final Long imageId) {
...
+            .viewType("FACE")
+            .imageType("OFF_BKG")
+            .active(true)
+            .sourceCode("GEN")
+            .offenderBooking(latestBooking.get())
+            .thumbnailImage(fullSizeImage ? null : Base64.getDecoder().decode(imageData))
+            .fullSizeImage(fullSizeImage ? Base64.getDecoder().decode(imageData) : null)
+            .build();
+
+        final OffenderImage savedImage = offenderImageRepository.save(newImage);
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:84]: Approach is often now to use JPA `addImage` on the OffenderBooking record ```    @OneToMany(mappedBy = "offenderBooking", cascade = CascadeType.ALL)
    @Default
    @Exclude
    @BatchSize(size = 25)
    private List<OffenderImage> images = new ArrayList<>();```

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java [hunk 13]
```diff
@@ -65,4 +67,65 @@ public void getImageDataReturnsNotFound() {
...
+        assertThat(responseEntity.getBody()).isInstanceOf(ImageDetail.class);
+        final var imageDetail = responseEntity.getBody();
+        assertThat(imageDetail).isNotNull();
+        assertThat(imageDetail.getImageId()).isGreaterThan(0);
+        assertThat(imageDetail.getCaptureDate()).isAfter(LocalDate.now().minusDays(1));
+        assertThat(imageDetail.getImageType()).isEqualTo("OFF_BKG");
+    }
+
+    @Test
+    public void putImageUploadsFullSizeImage() {
```
Comment (Mjwillis) [inline src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java:115]: Could have a test for a user not in the correct prison

**Review by Mjwillis — COMMENTED** [review]
See comments

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java (hunk 4 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:77]: Removed

File: README.MD (hunk 2 above)
Comment (timharrison-moj) [reply README.MD:223]: Done.

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java (hunk 5 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:81]: Done.

File: README.MD (hunk 3 above)
Comment (timharrison-moj) [reply README.MD:225]: Does Nomis use standard sizes for the full and thumbnail images to get around differences in cameras used at different sites?

Comment (timharrison-moj) [reply README.MD:225]: I could get the original size, and if larger scale down for the fullsize stored image, and scale down again for the thumbnail?

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java (hunk 7 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java:3]: Just removing additional blank line in passing.

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java (hunk 8 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:54]: Done

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java (hunk 9 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:57]: Added

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java (hunk 10 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:59]: Done

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java (hunk 11 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:65]: Done

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java (hunk 6 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:86]: It could, but is there much difference? Client is required to base64-encode the image and String might be a more universal type?

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java (hunk 13 above)
Comment (timharrison-moj) [reply src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java:115]: I was thinking this was a purely administrative function rather than something to build into a front-end with real prison users. I see why the byte-array type question on the API now, to upload a file directly?

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java (hunk 12 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:84]: Altered to use this method on the booking entity.
`public ImageDetail addImage(OffenderImage image) {
        getImages().add(image);
        image.setOffenderBooking(this);
        return image.transform();
  }`

File: README.MD (hunk 3 above)
Comment (timharrison-moj) [reply README.MD:225]: Implemented a scaling method, so going to find what the standard full and thumbnail image sizes are, and implement those.
`private byte[] scaleImage(int width, int height, byte[] source) throws IOException, IllegalArgumentException {
        InputStream is = new ByteArrayInputStream(source);
        BufferedImage bin = ImageIO.read(is);
        BufferedImage bout = new BufferedImage(width, height, bin.getType());
        Graphics2D g2d = bout.createGraphics();
        g2d.drawImage(bin, 0, 0, width, height, null);
        g2d.dispose();
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ImageIO.write(bout, "jpg", baos);
        return baos.toByteArray();
    }`

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java (hunk 13 above)
Comment (timharrison-moj) [reply src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java:115]: But yes,  I can do some test to verify.

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java (hunk 12 above)
Comment (timharrison-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:84]: This does insert the image correctly, but does not populate the imageId for the response back as no specific `saveAndFlush()`. Would be nice to get the imageId for the response.

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java (hunk 13 above)
Comment (timharrison-moj) [reply src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ImageResourceIntTest.java]: Actually, I'm not sure how to do this? Integration tests use ITAG_USER, which will have a caseload setup, but are there offenders in the int. test data to whom ITAG_USER does not have access?

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java [hunk 14]
```diff
@@ -17,7 +16,7 @@
     List<OffenderImage> getImagesByOffenderNumber(@Param("offenderNumber") final String offenderNumber);
 
     @Query(value = "select oi from OffenderImage oi where oi.id = (SELECT max(i.id) FROM OffenderImage i join i.offenderBooking ob join ob.offender o where o.nomsId = :offenderNumber " +
-        "and i.viewType = 'FACE' and i.orientationType = 'FRONT' and i.imageType = 'OFF_BKG' and i.active = true)")
+        "and ob.bookingSequence = 1 and i.viewType = 'FACE' and i.orientationType = 'FRONT' and i.imageType = 'OFF_BKG' and i.active = true)")
     Optional<OffenderImage> findLatestByOffenderNumber(@Param("offenderNumber") final String offenderNumber);
```
Comment (timharrison-moj) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java:20]: This (I think) fixes the issue where old images (from old bookings) were being returned when accessing images by `offenderNumber`. Accessing by `bookingId` would work ok.

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java [hunk 15]
```diff
@@ -66,4 +72,17 @@
...
+
+    @ApiResponses({
+        @ApiResponse(code = 400, message = "Invalid request.", response = ErrorResponse.class),
+        @ApiResponse(code = 404, message = "The offender number could not be found or has no bookings.", response = ErrorResponse.class),
+        @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class)})
+    @ApiOperation(value = "Upload a new image for an offender.", notes = "Image data is base64-encoded JPEG content. Requires ROLE_IMAGE_UPLOAD.", nickname = "putImage", hidden = true)
+    @PutMapping("/offender/{offenderNo}")
+    public ImageDetail putImage(
+        @Pattern(regexp = "^[A-Z]\\d{4}[A-Z]{2}$", message = "Offender Number format incorrect") @PathVariable("offenderNo") @ApiParam(value = "The offender number relating to this image.", required = true) final String offenderNo,
+        @RequestBody @ApiParam(value = "The base64-encoded string of the (full-sized) JPEG image data.", required = true) final String imageData
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/ImageResource.java:84]: Think we'll do another PR for changing this to multipart.

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java (hunk 14 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderImageRepository.java:20]: Suspect in prod its not an issue

File: src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java [hunk 16]
```diff
@@ -39,13 +52,94 @@ public ImageDetail findImageDetail(final Long imageId) {
...
+        final var offender = offenderRepository
+            .findOffenderByNomsId(offenderNumber)
+            .orElseThrow(EntityNotFoundException.withMessage(format("No prisoner found for prisoner number %s", offenderNumber)));
+
+        var booking = offender
+            .getLatestBooking()
+            .orElseThrow(EntityNotFoundException.withMessage(format("There are no bookings for %s", offenderNumber)));
+
+        // Set the previously active facial image for this bookingId to inactive
+        var previousImage = offenderImageRepository.findLatestByBookingId(booking.getBookingId());
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/service/ImageService.java:81]: Could have used model to set previous to inactive as already loaded OffenderBooking object.  Possible future change

**Review by Mjwillis — APPROVED** [review]
LGTM
